### PR TITLE
Adds a custom MarshalJSON method for trace.Link

### LINKED ISSUE #1820: trace.Link omits its Attributes field when marshalled to JSON
### Description

Because `trace.Link` embeds `trace.SpanContext` and `trace.SpanContext` has a custom `MarshalJSON` method, `trace.Link`'s other fields, i.e. `Attributes` do not get marshalled.

### Environment

- OS: Windows
- Architecture: amd64
- Go Version: 1.16.3
- opentelemetry-go version: v0.19.0

### Steps To Reproduce

```
json.Marshal(trace.Link{spanContext, []attribute.KeyValue{attribute.String("key", "value")})
```

### Expected behavior

JSON marshalling should include the attributes of the link.

### PR DESCRIPTION
**Pull request by IrisTuntun**
Adds a custom MarshalJSON method for `trace.Link` to marshal all fields.
Should resolve #1820

### CODE REVIEW

**Review by MrAlias — COMMENTED** [review]
Should we just not embed the `SpanContext` in a `Link`? Instead make it a field?

**Comment by IrisTuntun** [comment]
> Should we just not embed the `SpanContext` in a `Link`? Instead make it a field?

Then it will be another issue. I wonder if I should make the change in this PR.

**Comment by MrAlias** [comment]
> Then it will be another issue. I wonder if I should make the change in this PR.

I would get the opinion of another approver/maintainer before restarting.

File: CHANGELOG.md [hunk 1]
```diff
@@ -9,14 +9,48 @@ This project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.htm
 ## [Unreleased]
 
 ### Added
+- Adds a custom MarshalJSON method for `trace.Link` to marshal all fields. (#1852)
 
-- - Extract resource attributes from the `OTEL_RESOURCE_ATTRIBUTES` environment variable and merge them with the `resource.Default` resource as well as resources provided to the `TracerProvider` and metric `Controller`. (#1785)
```
Comment (MrAlias) [inline CHANGELOG.md:13]: This CHANGELOG is not properly merged. It needs to be sync-ed with main and only contain the changes relevant to this PR.

**Review by MrAlias — CHANGES_REQUESTED** [review]

**Comment by Aneurysm9** [comment]
> Should we just not embed the `SpanContext` in a `Link`? Instead make it a field?

That would also work.  I don't have any context for why `SpanContext` was embedded rather than included as a field, but I'd presume that it was to avoid the need to write passthrough methods to get at the linked trace and span IDs.  That said, it doesn't look like the spec provides for any sort of interface for interacting with `Link`s and API `Span`s are write-only, so once they're added there's no way to access their properties through the API.

As I was reviewing #1846 I was wondering whether the `Link` type should be moved as well.  It would be slightly more complicated since the API does have a `Span#AddLink()` method, but the spec certainly contemplates that taking separate `SpanContext` and `Attributes` arguments, so we could restructure.

**Comment by Aneurysm9** [comment]
I see I didn't look closely enough at the issue #1846 was addressing and this topic is discussed there.  Added some thoughts: https://github.com/open-telemetry/opentelemetry-go/issues/1845#issuecomment-829487787.

**Comment by MrAlias** [comment]
Closing as the decision outlined in https://github.com/open-telemetry/opentelemetry-go/issues/1845#issuecomment-829594030 is to remove the `Link` type from the API. We can recreate the type in the SDK without embedding the `SpanContext`.

**Comment by MrAlias** [comment]
@IrisTuntun we realized that we cannot remove the link and we should instead go with the un-embed solution: https://github.com/open-telemetry/opentelemetry-go/issues/1845#issuecomment-829700761

Is this something you would like to open a PR for?

**Comment by IrisTuntun** [comment]
> @IrisTuntun we realized that we cannot remove the link and we should instead go with the un-embed solution: [#1845 (comment)](https://github.com/open-telemetry/opentelemetry-go/issues/1845#issuecomment-829700761)
> 
> Is this something you would like to open a PR for?

Sure, I can take this. Will open a PR for it.

**Pull request closed without merging**